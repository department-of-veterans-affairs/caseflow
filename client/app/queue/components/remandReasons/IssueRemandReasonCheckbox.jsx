import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';

import { css } from 'glamor';

import Checkbox from 'app/components/Checkbox';
import RadioField from 'app/components/RadioField';
import COPY from 'app/../COPY';
import {
  errorNoTopMargin,
  smallBottomMargin,
  smallLeftMargin,
} from './constants';

export const IssueRemandReasonCheckbox = ({
  isLegacyAppeal = false,
  highlight = false,
  option = {},
  onChange,
  prefix = '',
  value: valueProp,
}) => {
  const [state, setState] = useState({ value: false, post_aoj: null });
  const firstUpdate = useRef(true);
  const copyPrefix = isLegacyAppeal ? 'LEGACY' : 'AMA';

  const handleCheckboxChange = (value) => setState({ value, post_aoj: null });
  const handleRadioChange = (val) => setState({ ...state, post_aoj: val });

  useEffect(() => {
    // No need to hit callback until something changes
    if (firstUpdate.current) {
      firstUpdate.current = false;

      return;
    }

    onChange?.({
      code: option.id,
      checked: state.value,
      post_aoj: state.post_aoj,
    });
  }, [state.value, state.post_aoj]);

  useEffect(() => {
    if (valueProp) {
      setState({
        value: valueProp.checked ?? true,
        // eslint-disable-next-line camelcase
        post_aoj: valueProp?.post_aoj?.toString() ?? null,
      });
    }
  }, [valueProp]);

  const prefixedId = `${prefix}-${option.id}`;

  return (
    <React.Fragment key={prefixedId}>
      <Checkbox
        name={prefixedId}
        onChange={handleCheckboxChange}
        value={state.value}
        label={option.label}
        unpadded
      />
      {state.value && (
        <RadioField
          // eslint-disable-next-line no-undefined
          errorMessage={(highlight && state.post_aoj === null) ? 'Choose one' : undefined}
          styling={css(smallLeftMargin, smallBottomMargin, errorNoTopMargin)}
          name={`${prefixedId}-postAoj`}
          vertical
          hideLabel
          options={[
            {
              displayText:
                COPY[`${copyPrefix}_REMAND_REASON_POST_AOJ_LABEL_BEFORE`],
              value: 'false',
            },
            {
              displayText:
                COPY[`${copyPrefix}_REMAND_REASON_POST_AOJ_LABEL_AFTER`],
              value: 'true',
            },
          ]}
          value={String(state.post_aoj)}
          onChange={handleRadioChange}
        />
      )}
    </React.Fragment>
  );
};
IssueRemandReasonCheckbox.propTypes = {
  isLegacyAppeal: PropTypes.bool,
  highlight: PropTypes.bool,
  option: PropTypes.shape({
    id: PropTypes.string,
    label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  }),
  onChange: PropTypes.func,
  prefix: PropTypes.string,
  value: PropTypes.shape({
    code: PropTypes.string.isRequired,
    checked: PropTypes.bool.isRequired,
    post_aoj: PropTypes.bool.isRequired,
  }),
};
