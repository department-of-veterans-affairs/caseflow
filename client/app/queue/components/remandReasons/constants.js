import { css } from 'glamor';

export const smallLeftMargin = css({ marginLeft: '1rem' });
export const smallBottomMargin = css({ marginBottom: '1rem' });
export const errorNoTopMargin = css({
  '.usa-input-error': { marginTop: 0 },
});
export const flexContainer = css({
  display: 'flex',
  maxWidth: '75rem',
});
export const flexColumn = css({
  flexDirection: 'row',
  flexWrap: 'wrap',
  width: '50%',
});
