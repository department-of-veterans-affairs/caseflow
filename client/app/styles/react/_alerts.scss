
.usa-alert {

  &.fixed {
    position: fixed;
    top: 0;
    margin-top: 0;
    z-index: 999;
    left: 50%;
    transform: translate(-50%, 0);
  }

  // Remove after USWDS V.2.0
  &.usa-alert-slim {
    background-size: 4rem;

    .usa-alert-body {
      padding-left: 3.5rem;
    }

    .usa-alert-text {
      font-size: 2rem;
    }
  }

  .usa-alert-text {
    line-height: 1.5;
  }
}
