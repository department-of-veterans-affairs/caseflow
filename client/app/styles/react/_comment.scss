.comment-control-button-container {
  display: flex;
  width: 100%;
  padding: 3px 0;
  justify-content: space-between;
}

.comment-save-button-container {
  display: inline-block;
  width: 100%;
  margin-top: 5px;
  margin-bottom: 17px;
}

.comment-control-button {
  padding: 0;

  &.horizontal {
    text-align: left;
  }
}

.comment-control-button-divider {
  padding: 0 10px;
}

.comment-container-selected {
  box-shadow: 3px 3px 5px $color-gray-light;
}

.comment-container {
  word-wrap: break-word;
  word-break: break-word;
  background-color: $color-white;
  padding: 15px;
  cursor: pointer;

  &:not(.comment-horizontal-container) {
    margin-bottom: 15px;
    border-radius: 3px;
    border: 1px solid $color-gray-lighter;
  }

  &.comment-horizontal-container {
    padding: 25px 25px 25px 40px;
  }
}

// Max width is enforced so the size of this comment box
// fits with existing textareas in style guide
.comment-size-container {
  max-width: 430px;

  .cf-form-textarea {
    margin-bottom: 20px;
  }
}

.comment-horizontal-container {
  display: inline-block;
  width: 100%;
}

.comment-horizontal-spacing {
  margin-right: 49px;
}

.commentIcon-container {
  position: absolute;
  z-index: 10;
  // These dimensions need to match the comment icon SVG dimensions.
  width: 40px;
  height: 40px;
}

