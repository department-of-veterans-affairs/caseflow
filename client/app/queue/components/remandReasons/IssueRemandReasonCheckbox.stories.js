import React from 'react';

import { IssueRemandReasonCheckbox } from './IssueRemandReasonCheckbox';

/* eslint-disable react/prop-types */

const option = { id: 'lorem', label: 'Lorem Ipsum' };

export default {
  title: 'Queue/Components/Remand Reasons/IssueRemandReasonCheckbox',
  component: IssueRemandReasonCheckbox,
  parameters: {
    controls: { expanded: true },
  },
  args: {
    option,
  },
  argTypes: {
    onChange: { action: 'onChange' },
  },
};

const Template = (args) => <IssueRemandReasonCheckbox {...args} />;

export const AMA = Template.bind({});
AMA.args = { isLegacyAppeal: false };

export const Legacy = Template.bind({});
Legacy.args = { isLegacyAppeal: true };

export const DefaultValue = Template.bind({});
DefaultValue.args = {
  isLegacyAppeal: false,
  value: { code: option.id, post_aoj: true },
};
