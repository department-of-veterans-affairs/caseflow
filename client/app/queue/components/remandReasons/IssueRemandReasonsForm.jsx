import React, { useEffect, useMemo, useRef, useState } from 'react';
import PropTypes from 'prop-types';

import { css } from 'glamor';

import CheckboxGroup from 'app/components/CheckboxGroup';
import {
  LEGACY_REMAND_REASONS,
  REMAND_REASONS,
  fullWidth,
  boldText,
  redText,
} from 'app/queue/constants';
import BENEFIT_TYPES from 'constants/BENEFIT_TYPES';
import { IssueRemandReasonCheckbox } from './IssueRemandReasonCheckbox';
import {
  errorNoTopMargin,
  flexColumn,
  flexContainer,
  smallBottomMargin,
} from './constants';
import {
  getIssueDiagnosticCodeLabel,
  getIssueProgramDescription,
  getIssueTypeDescription,
} from '../../utils';
import { formatDateStr } from '../../../util/DateUtil';

export const LegacyCheckboxGroup = ({
  onChange,
  highlight,
  prefix = '',
  values = {},
}) => {
  const getCheckbox = (option, checkboxChange) => (
    <IssueRemandReasonCheckbox
      option={option}
      onChange={checkboxChange}
      isLegacyAppeal
      highlight={highlight}
      prefix={prefix}
      value={values[option.id]}
    />
  );
  const checkboxGroupProps = {
    onChange,
    getCheckbox,
  };

  return (
    <div {...flexContainer}>
      <div {...flexColumn}>
        <CheckboxGroup
          label={<h3>Medical examination and opinion</h3>}
          name="med-exam"
          options={LEGACY_REMAND_REASONS.medicalExam}
          {...checkboxGroupProps}
        />
        <CheckboxGroup
          label={<h3>Duty to assist records request</h3>}
          name="duty-to-assist"
          options={LEGACY_REMAND_REASONS.dutyToAssistRecordsRequest}
          {...checkboxGroupProps}
        />
      </div>
      <div {...flexColumn}>
        <CheckboxGroup
          label={<h3>Duty to notify</h3>}
          name="duty-to-notify"
          options={LEGACY_REMAND_REASONS.dutyToNotify}
          {...checkboxGroupProps}
        />
        <CheckboxGroup
          label={<h3>Due process</h3>}
          name="due-process"
          options={LEGACY_REMAND_REASONS.dueProcess}
          {...checkboxGroupProps}
        />
      </div>
    </div>
  );
};
LegacyCheckboxGroup.propTypes = {
  highlight: PropTypes.bool,
  onChange: PropTypes.func,
  prefix: PropTypes.string,
  values: PropTypes.shape({
    [PropTypes.string]: PropTypes.shape({
      code: PropTypes.string.isRequired,
      checked: PropTypes.bool,
      post_aoj: PropTypes.bool.isRequired,
    }),
  }),
};

export const AmaCheckboxGroup = ({
  onChange,
  highlight,
  prefix = '',
  values = {},
}) => {
  const getCheckbox = (option, checkboxChange) => (
    <IssueRemandReasonCheckbox
      option={option}
      onChange={checkboxChange}
      highlight={highlight}
      prefix={prefix}
      value={values[option.id]}
    />
  );
  const checkboxGroupProps = {
    onChange,
    getCheckbox,
  };

  return (
    <div {...flexContainer}>
      <div {...flexColumn}>
        <CheckboxGroup
          label={<h3>Duty to notify</h3>}
          name="duty-to-notify"
          options={REMAND_REASONS.dutyToNotify}
          {...checkboxGroupProps}
        />
        <CheckboxGroup
          label={<h3>Duty to assist</h3>}
          name="duty-to-assist"
          options={REMAND_REASONS.dutyToAssist}
          {...checkboxGroupProps}
        />
      </div>
      <div {...flexColumn}>
        <CheckboxGroup
          label={<h3>Medical examination</h3>}
          name="medical-exam"
          options={REMAND_REASONS.medicalExam}
          {...checkboxGroupProps}
        />
        <br />
        <CheckboxGroup
          label={<h3>Due Process</h3>}
          name="due-process"
          options={REMAND_REASONS.dueProcess}
          {...checkboxGroupProps}
        />
      </div>
    </div>
  );
};
AmaCheckboxGroup.propTypes = {
  highlight: PropTypes.bool,
  onChange: PropTypes.func,
  prefix: PropTypes.string,
  values: PropTypes.shape({
    [PropTypes.string]: PropTypes.shape({
      code: PropTypes.string.isRequired,
      checked: PropTypes.bool,
      post_aoj: PropTypes.bool.isRequired,
    }),
  }),
};

export const IssueRemandReasonsForm = ({
  certificationDate,
  isLegacyAppeal = false,
  issue,
  issueNumber = 1,
  issueTotal = 1,
  highlight,
  onChange,
  values = [],
}) => {
  const [fields, setFields] = useState({});
  const firstUpdate = useRef(true);

  // Reformat initial/default values
  useEffect(() => {
    if (!values.length) {
      return;
    }

    const newFields = {};

    values.forEach(
      // eslint-disable-next-line camelcase
      ({ code, post_aoj }) =>
        (newFields[code] = { code, checked: true, post_aoj })
    );
    setFields(newFields);
  }, [values]);

  // eslint-disable-next-line camelcase
  const handleChange = ({ code, checked, post_aoj }) => {
    setFields({
      ...fields,
      [code]: { code, checked, post_aoj },
    });
  };
  const checkboxGroupProps = {
    onChange: handleChange,
    highlight,
    prefix: `issue-${issue.id}`,
    values: fields,
  };

  // Memoized array of the selected options; used for onChange and validation
  const selected = useMemo(
    () => Object.values(fields).filter((item) => item.checked),
    [fields]
  );

  useEffect(() => {
    // No need to hit callback until something changes
    if (firstUpdate.current) {
      firstUpdate.current = false;

      return;
    }

    onChange?.(selected);
  }, [selected]);

  return (
    <div
      className="remand-reasons-form"
      key={`remand-reasons-${String(issue.id)}`}
    >
      <h2 className="cf-push-left" {...css(fullWidth, smallBottomMargin)}>
        Issue {issueNumber} {issueTotal > 1 ? ` of ${issueTotal}` : ''}
      </h2>
      <div {...smallBottomMargin}>
        {isLegacyAppeal ?
          `Program: ${getIssueProgramDescription(issue)}` :
          `Benefit type: ${BENEFIT_TYPES[issue.benefit_type]}`}
      </div>
      {!isLegacyAppeal && (
        <div {...smallBottomMargin}>Issue description: {issue.description}</div>
      )}
      {isLegacyAppeal && (
        <React.Fragment>
          <div {...smallBottomMargin}>
            Issue: {getIssueTypeDescription(issue)}
          </div>
          <div {...smallBottomMargin}>
            Code:{' '}
            {getIssueDiagnosticCodeLabel(issue.codes[issue.codes.length - 1])}
          </div>
          <div
            {...smallBottomMargin}
            ref={(node) => (this.elTopOfWarning = node)}
          >
            Certified: {formatDateStr(certificationDate)}
          </div>
          <div {...smallBottomMargin}>Note: {issue.note}</div>
        </React.Fragment>
      )}
      {highlight && !selected.length && (
        <div
          className="usa-input-error"
          {...css(redText, boldText, errorNoTopMargin)}
        >
          Choose at least one
        </div>
      )}
      {isLegacyAppeal ? (
        <LegacyCheckboxGroup {...checkboxGroupProps} />
      ) : (
        <AmaCheckboxGroup {...checkboxGroupProps} />
      )}
    </div>
  );
};
IssueRemandReasonsForm.propTypes = {
  certificationDate: PropTypes.string,
  isLegacyAppeal: PropTypes.bool,
  issue: PropTypes.object.isRequired,
  issueNumber: PropTypes.number,
  issueTotal: PropTypes.number,
  highlight: PropTypes.bool,
  onChange: PropTypes.func,
  values: PropTypes.arrayOf(
    PropTypes.shape({
      code: PropTypes.string,
      checked: PropTypes.bool,
      post_aoj: PropTypes.bool,
    })
  ),
};
