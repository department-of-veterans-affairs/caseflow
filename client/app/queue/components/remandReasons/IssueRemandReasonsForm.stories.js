import React from 'react';

import { IssueRemandReasonsForm } from './IssueRemandReasonsForm';

/* eslint-disable react/prop-types */

const issue = {
  id: 1,
  benefit_type: 'education',
  description: 'Lorem ipsum and whatnot',
  diagnostic_code: '503',
  disposition: 'remanded',
};

export default {
  title: 'Queue/Components/Remand Reasons/IssueRemandReasonsForm',
  component: IssueRemandReasonsForm,
  parameters: {
    controls: { expanded: true },
  },
  args: {
    issue,
    issueNumber: 1,
    issueTotal: 2,
  },
  argTypes: {
    onChange: { action: 'onChange' },
  },
};

const Template = (args) => <IssueRemandReasonsForm {...args} />;

export const AMA = Template.bind({});
AMA.args = { isLegacyAppeal: false };

export const Legacy = Template.bind({});
Legacy.args = { isLegacyAppeal: true };

export const DefaultValues = Template.bind({});
DefaultValues.args = {
  isLegacyAppeal: false,
  values: [
    { code: 'incorrect_notice_sent', checked: true, post_aoj: false },
    { code: 'medical_opinions', checked: true, post_aoj: true },
  ]
};
