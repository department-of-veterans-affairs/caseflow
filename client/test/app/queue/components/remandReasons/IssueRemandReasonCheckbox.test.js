import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { IssueRemandReasonCheckbox } from 'app/queue/components/remandReasons/IssueRemandReasonCheckbox';
import COPY from 'COPY';

describe('IssueRemandReasonCheckbox', () => {
  const onChange = jest.fn();
  const option = { id: 'lorem', label: 'Lorem Ipsum' };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('AMA', () => {
    it('renders correctly by default', () => {
      const { container } = render(
        <IssueRemandReasonCheckbox option={option} />
      );

      expect(container).toMatchSnapshot();

      expect(screen.getByText(option.label)).toBeInTheDocument();
    });

    it('hides AOJ choice until checked', async () => {
      const { container } = render(
        <IssueRemandReasonCheckbox option={option} />
      );

      expect(
        screen.queryByText(COPY.AMA_REMAND_REASON_POST_AOJ_LABEL_BEFORE)
      ).not.toBeTruthy();
      expect(
        screen.queryByText(COPY.AMA_REMAND_REASON_POST_AOJ_LABEL_AFTER)
      ).not.toBeTruthy();

      userEvent.click(screen.getByRole('checkbox'));

      expect(
        screen.queryByText(COPY.AMA_REMAND_REASON_POST_AOJ_LABEL_BEFORE)
      ).toBeTruthy();
      expect(
        screen.queryByText(COPY.AMA_REMAND_REASON_POST_AOJ_LABEL_AFTER)
      ).toBeTruthy();

      expect(container).toMatchSnapshot();
    });

    it('properly fires onChange', async () => {
      render(<IssueRemandReasonCheckbox option={option} onChange={onChange} />);

      userEvent.click(screen.getByRole('checkbox'));

      expect(onChange).toHaveBeenCalledTimes(1);
      expect(onChange).toHaveBeenLastCalledWith({
        code: option.id,
        checked: true,
        post_aoj: null,
      });

      userEvent.click(
        screen.getByLabelText(COPY.AMA_REMAND_REASON_POST_AOJ_LABEL_BEFORE)
      );

      expect(onChange).toHaveBeenCalledTimes(2);
      expect(onChange).toHaveBeenLastCalledWith({
        code: option.id,
        checked: true,
        post_aoj: 'false',
      });

      userEvent.click(
        screen.getByLabelText(COPY.AMA_REMAND_REASON_POST_AOJ_LABEL_AFTER)
      );

      expect(onChange).toHaveBeenCalledTimes(3);
      expect(onChange).toHaveBeenLastCalledWith({
        code: option.id,
        checked: true,
        post_aoj: 'true',
      });
    });
  });
});
