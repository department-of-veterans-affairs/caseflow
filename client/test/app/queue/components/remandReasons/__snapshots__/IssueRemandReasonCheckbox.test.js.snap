// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssueRemandReasonCheckbox AMA hides AOJ choice until checked 1`] = `
<div>
  <div
    class="checkbox-wrapper--lorem"
  >
    <div
      class="cf-form-checkbox"
    >
      <input
        aria-label="-lorem"
        id="-lorem"
        name="-lorem"
        type="checkbox"
      />
      <label
        for="-lorem"
      >
        <span
          class=""
        >
          Lorem Ipsum
        </span>
         
      </label>
    </div>
  </div>
  <fieldset
    class="usa-fieldset-inputs cf-form-radio "
    data-css-10hr1g7=""
  >
    <legend
      class="usa-sr-only"
    >
      <span>
        -lorem-postAoj
         
      </span>
    </legend>
    <div
      class="cf-form-radio-options"
    >
      <div
        class="cf-form-radio-option"
      >
        <input
          id="-lorem-postAoj_false"
          name="-lorem-postAoj"
          type="radio"
          value="false"
        />
        <label
          class=""
          for="-lorem-postAoj_false"
        >
          Pre AOJ
        </label>
      </div>
      <div
        class="cf-form-radio-option"
      >
        <input
          id="-lorem-postAoj_true"
          name="-lorem-postAoj"
          type="radio"
          value="true"
        />
        <label
          class=""
          for="-lorem-postAoj_true"
        >
          Post AOJ
        </label>
      </div>
    </div>
  </fieldset>
</div>
`;

exports[`IssueRemandReasonCheckbox AMA renders correctly by default 1`] = `
<div>
  <div
    class="checkbox-wrapper--lorem"
  >
    <div
      class="cf-form-checkbox"
    >
      <input
        aria-label="-lorem"
        id="-lorem"
        name="-lorem"
        type="checkbox"
      />
      <label
        for="-lorem"
      >
        <span
          class=""
        >
          Lorem Ipsum
        </span>
         
      </label>
    </div>
  </div>
</div>
`;
